File: mt5-core-server/src/services/auto/analyzers/smc/marketStructure.ts
/**
 * SMC Engine V2 — Market Structure Detection
 * Port จาก Pine Script V8.3: BOS / CHoCH / SMS / BMS + IDM Filter
 *
 * ตรวจจับ:
 * - Break of Structure (BOS) — continuation
 * - Change of Character (CHoCH) — reversal
 * - SMS (Shift in Market Structure) — first continuation after CHoCH
 * - BMS (Break in Market Structure) — subsequent continuations
 * - IDM (Inducement) Filter — require liquidity sweep before valid break
 *
 * 2026-05-06 — skyliner.jojo@gmail.com
 */

import type { Candle, MarketStructure, StructureDirection, StructureEvent } from './types.js';
import { detectSwings } from './utils.js';

export interface MarketStructureConfig {
  useBodyBreak: boolean;      // ใช้ body หรือ wick ในการเช็ค break (default: true = body)
  swingLength: number;        // swing detection lookback (default: 10)
  idmEnabled: boolean;        // เปิดใช้ IDM filter (default: true)
  idmSwingLength: number;     // IDM swing length (default: 3)
  idmApplyTo: 'CHoCH' | 'CHoCH+BOS'; // default: 'CHoCH'
}

const DEFAULT_CONFIG: MarketStructureConfig = {
  useBodyBreak: true,
  swingLength: 10,
  idmEnabled: true,
  idmSwingLength: 3,
  idmApplyTo: 'CHoCH',
};

/**
 * Detect Market Structure from candle array.
 * State-based approach matching Pine Script V8.3 logic.
 */
export function detectMarketStructure(
  candles: Candle[],
  config: Partial<MarketStructureConfig> = {}
): MarketStructure {
  const cfg = { ...DEFAULT_CONFIG, ...config };

  if (candles.length < 30) {
    return defaultStructure();
  }

  // State variables (equivalent to Pine Script `var`)
  let direction: StructureDirection = 'INIT';
  let structureHigh = candles[0].h;
  let structureLow = candles[0].l;
  let structureHighBarIndex = 0;
  let structureLowBarIndex = 0;
  let lastEvent: StructureEvent = 'NONE';
  let lastEventSide: 'UP' | 'DOWN' | null = null;
  let continuationCount = 0;

  // IDM state
  let idmHigh: number | null = null;
  let idmLow: number | null = null;
  let idmHighSwept = false;
  let idmLowSwept = false;

  // Detect swing points for IDM
  const idmSwings = cfg.idmEnabled ? detectSwings(candles, cfg.idmSwingLength) : { highs: [], lows: [] };

  // Process each bar
  for (let i = 1; i < candles.length; i++) {
    const c = candles[i];
    const breakLow = cfg.useBodyBreak ? c.c : c.l;
    const breakHigh = cfg.useBodyBreak ? c.c : c.h;

    // --- IDM Processing ---
    if (cfg.idmEnabled) {
      // Update IDM swing points
      for (const sh of idmSwings.highs) {
        if (sh.index === i) {
          idmHigh = sh.price;
          idmHighSwept = false;
        }
      }
      for (const sl of idmSwings.lows) {
        if (sl.index === i) {
          idmLow = sl.price;
          idmLowSwept = false;
        }
      }

      // Check IDM sweeps (wick through but close back)
      if (!idmHighSwept && idmHigh !== null) {
        if (c.h > idmHigh && c.c < idmHigh) idmHighSwept = true;
        if (i > 0 && candles[i - 1].h > idmHigh && candles[i - 1].c < idmHigh) idmHighSwept = true;
      }
      if (!idmLowSwept && idmLow !== null) {
        if (c.l < idmLow && c.c > idmLow) idmLowSwept = true;
        if (i > 0 && candles[i - 1].l < idmLow && candles[i - 1].c > idmLow) idmLowSwept = true;
      }
    }

    // --- Dynamic Lookback for Structure Extremes ---
    const lbHigh = Math.max(1, Math.min(i - structureLowBarIndex, i));
    const lbLow = Math.max(1, Math.min(i - structureHighBarIndex, i));

    // Find highest high since last structure low
    let highestInRange = structureHigh;
    let highestIdx = structureHighBarIndex;
    for (let j = Math.max(0, i - lbHigh); j <= i; j++) {
      if (candles[j].h > highestInRange) {
        highestInRange = candles[j].h;
        highestIdx = j;
      }
    }

    // Find lowest low since last structure high
    let lowestInRange = structureLow;
    let lowestIdx = structureLowBarIndex;
    for (let j = Math.max(0, i - lbLow); j <= i; j++) {
      if (candles[j].l < lowestInRange) {
        lowestInRange = candles[j].l;
        lowestIdx = j;
      }
    }

    // --- Break Detection ---
    const isLowBroken = breakLow < structureLow && i > structureLowBarIndex;
    const isHighBroken = breakHigh > structureHigh && i > structureHighBarIndex;

    if (isLowBroken) {
      const priorDir = direction;
      const isReversal = priorDir !== 'BEARISH';

      // IDM validation
      const needIDM = cfg.idmEnabled && (
        (cfg.idmApplyTo === 'CHoCH' && isReversal) ||
        cfg.idmApplyTo === 'CHoCH+BOS'
      );
      const idmOk = !needIDM || idmHighSwept;

      if (idmOk) {
        if (isReversal) {
          lastEvent = 'CHoCH';
          continuationCount = 0;
        } else {
          continuationCount++;
          lastEvent = continuationCount === 1 ? 'SMS' : 'BMS';
        }
        lastEventSide = 'DOWN';
        direction = 'BEARISH';

        // Update extremes
        structureHighBarIndex = highestIdx;
        structureHigh = highestInRange;
        structureLowBarIndex = i;
        structureLow = c.l;
      }
    } else if (isHighBroken) {
      const priorDir = direction;
      const isReversal = priorDir !== 'BULLISH';

      const needIDM = cfg.idmEnabled && (
        (cfg.idmApplyTo === 'CHoCH' && isReversal) ||
        cfg.idmApplyTo === 'CHoCH+BOS'
      );
      const idmOk = !needIDM || idmLowSwept;

      if (idmOk) {
        if (isReversal) {
          lastEvent = 'CHoCH';
          continuationCount = 0;
        } else {
          continuationCount++;
          lastEvent = continuationCount === 1 ? 'SMS' : 'BMS';
        }
        lastEventSide = 'UP';
        direction = 'BULLISH';

        // Update extremes
        structureLowBarIndex = lowestIdx;
        structureLow = lowestInRange;
        structureHighBarIndex = i;
        structureHigh = c.h;
      }
    } else {
      // Trail the extreme points
      if (direction === 'BULLISH') {
        if (c.h > structureHigh) {
          structureHigh = c.h;
          structureHighBarIndex = i;
        }
      } else if (direction === 'BEARISH') {
        if (c.l < structureLow) {
          structureLow = c.l;
          structureLowBarIndex = i;
        }
      } else {
        // INIT
        if (c.h > structureHigh) {
          structureHigh = c.h;
          structureHighBarIndex = i;
        }
        if (c.l < structureLow) {
          structureLow = c.l;
          structureLowBarIndex = i;
        }
      }
    }
  }

  return {
    direction,
    structureHigh,
    structureLow,
    structureHighBarIndex,
    structureLowBarIndex,
    lastEvent,
    lastEventSide,
    continuationCount,
    idmHigh,
    idmLow,
    idmHighSwept,
    idmLowSwept,
  };
}

function defaultStructure(): MarketStructure {
  return {
    direction: 'INIT',
    structureHigh: 0,
    structureLow: 0,
    structureHighBarIndex: 0,
    structureLowBarIndex: 0,
    lastEvent: 'NONE',
    lastEventSide: null,
    continuationCount: 0,
    idmHigh: null,
    idmLow: null,
    idmHighSwept: false,
    idmLowSwept: false,
  };
}

File: composeApp/src/commonMain/kotlin/com/example/personalaibot/automation/smc/MarketStructure.kt
package com.example.personalaibot.automation.smc

import com.example.personalaibot.tools.trading.Candle
import kotlin.math.max
import kotlin.math.min

/**
 * Market Structure Detection — port จาก mt5-core-server marketStructure.ts (Pine V8.3)
 * ตรวจ BOS / CHoCH / SMS / BMS + IDM (Inducement) filter
 */
internal object MarketStructureDetector {

    fun detect(
        candles: List<Candle>,
        useBodyBreak: Boolean = true,
        idmEnabled: Boolean = true,
        idmSwingLength: Int = 3
    ): MarketStructure {
        if (candles.size < 30) return emptyStructure()

        var direction = StructureDirection.INIT
        var structureHigh = candles[0].high
        var structureLow = candles[0].low
        var structureHighBarIndex = 0
        var structureLowBarIndex = 0
        var lastEvent = StructureEvent.NONE
        var lastEventSide: String? = null
        var continuationCount = 0

        var idmHigh: Double? = null
        var idmLow: Double? = null
        var idmHighSwept = false
        var idmLowSwept = false

        val idmSwings = if (idmEnabled) detectSwings(candles, idmSwingLength) else (emptyList<SwingPoint>() to emptyList())

        for (i in 1 until candles.size) {
            val c = candles[i]
            val breakLow = if (useBodyBreak) c.close else c.low
            val breakHigh = if (useBodyBreak) c.close else c.high

            // --- IDM ---
            if (idmEnabled) {
                for (sh in idmSwings.first) {
                    if (sh.index == i) { idmHigh = sh.price; idmHighSwept = false }
                }
                for (sl in idmSwings.second) {
                    if (sl.index == i) { idmLow = sl.price; idmLowSwept = false }
                }
                val ih = idmHigh
                if (!idmHighSwept && ih != null) {
                    if (c.high > ih && c.close < ih) idmHighSwept = true
                    if (candles[i - 1].high > ih && candles[i - 1].close < ih) idmHighSwept = true
                }
                val il = idmLow
                if (!idmLowSwept && il != null) {
                    if (c.low < il && c.close > il) idmLowSwept = true
                    if (candles[i - 1].low < il && candles[i - 1].close > il) idmLowSwept = true
                }
            }

            // --- Dynamic lookback ---
            val lbHigh = max(1, min(i - structureLowBarIndex, i))
            val lbLow = max(1, min(i - structureHighBarIndex, i))

            var highestInRange = structureHigh
            var highestIdx = structureHighBarIndex
            for (j in max(0, i - lbHigh)..i) {
                if (candles[j].high > highestInRange) {
                    highestInRange = candles[j].high
                    highestIdx = j
                }
            }

            var lowestInRange = structureLow
            var lowestIdx = structureLowBarIndex
            for (j in max(0, i - lbLow)..i) {
                if (candles[j].low < lowestInRange) {
                    lowestInRange = candles[j].low
                    lowestIdx = j
                }
            }

            val isLowBroken = breakLow < structureLow && i > structureLowBarIndex
            val isHighBroken = breakHigh > structureHigh && i > structureHighBarIndex

            if (isLowBroken) {
                val isReversal = direction != StructureDirection.BEARISH
                // IDM validation (idmApplyTo = 'CHoCH' เท่านั้น — ตาม default ต้นฉบับ)
                val needIdm = idmEnabled && isReversal
                if (!needIdm || idmHighSwept) {
                    if (isReversal) {
                        lastEvent = StructureEvent.CHoCH
                        continuationCount = 0
                    } else {
                        continuationCount++
                        lastEvent = if (continuationCount == 1) StructureEvent.SMS else StructureEvent.BMS
                    }
                    lastEventSide = "DOWN"
                    direction = StructureDirection.BEARISH
                    structureHighBarIndex = highestIdx
                    structureHigh = highestInRange
                    structureLowBarIndex = i
                    structureLow = c.low
                }
            } else if (isHighBroken) {
                val isReversal = direction != StructureDirection.BULLISH
                val needIdm = idmEnabled && isReversal
                if (!needIdm || idmLowSwept) {
                    if (isReversal) {
                        lastEvent = StructureEvent.CHoCH
                        continuationCount = 0
                    } else {
                        continuationCount++
                        lastEvent = if (continuationCount == 1) StructureEvent.SMS else StructureEvent.BMS
                    }
                    lastEventSide = "UP"
                    direction = StructureDirection.BULLISH
                    structureLowBarIndex = lowestIdx
                    structureLow = lowestInRange
                    structureHighBarIndex = i
                    structureHigh = c.high
                }
            } else {
                // trail extremes
                when (direction) {
                    StructureDirection.BULLISH -> if (c.high > structureHigh) {
                        structureHigh = c.high; structureHighBarIndex = i
                    }
                    StructureDirection.BEARISH -> if (c.low < structureLow) {
                        structureLow = c.low; structureLowBarIndex = i
                    }
                    StructureDirection.INIT -> {
                        if (c.high > structureHigh) { structureHigh = c.high; structureHighBarIndex = i }
                        if (c.low < structureLow) { structureLow = c.low; structureLowBarIndex = i }
                    }
                }
            }
        }

        return MarketStructure(
            direction, structureHigh, structureLow,
            structureHighBarIndex, structureLowBarIndex,
            lastEvent, lastEventSide, continuationCount,
            idmHigh, idmLow, idmHighSwept, idmLowSwept
        )
    }
}
